### PR TITLE
Online screenshot diff/generation using CircleCI

### PR DESCRIPTION
**Pull request by ylemkimon**
Added CircleCI config for generating diff image and new screenshots online: [example: diffs and generated screenshots are available at `Artifacts`](https://circleci.com/gh/ylemkimon/KaTeX/13).

**Things to consider**
- [ ] Enable or migrate to CircleCI on the main project
- [ ] Run lint/flow/jest
- [ ] Upload built file and provide a preview/demo/sandbox for each build(PR and push): #1171

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    - checkout
+    - run:
+        name: Checkout submodule
+        command: |
+            git submodule sync
+            git submodule update --init --recursive
+    - run:
+        name: Checkout unicode fonts
+        command: |
+            git clone https://github.com/Khan/KaTeX-test-fonts test/screenshotter/unicode-fonts
```
Comment (ylemkimon) [inline .circleci/config.yml]: This can be removed if #1182 is merged.

**Comment by kevinbarabash** [comment]
@ylemkimon this looks really cool.  How do I view the generated screenshots?

**Comment by ylemkimon** [comment]
@kevinbarabash Diffs and generated screenshots are available at [`Artifacts`](https://circleci.com/gh/ylemkimon/KaTeX/13) in CircleCI build page.

**Comment by kevinbarabash** [comment]
I wasn't seeing the "Artifacts" tab b/c I wasn't logged in.  It's there now.  I'm surprised at how quickly the tests run compared to travis-ci.

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    - save_cache:
+        key: npm-deps-{{ checksum "package.json" }}
+        paths:
+          - node_modules
+
+    - run:
+        name: Verify screenshots and generate diff
+        command: node dockers/Screenshotter/screenshotter.js --seleniumIP localhost -b $CIRCLE_JOB --verify --diff
+    - run:
+        name: Generate new screenshots
```
Comment (kevinbarabash) [inline .circleci/config.yml]: So we're both generating and verifying the screenshots and it's faster than travis-ci.  Awesome!

File: .circleci/config.yml (hunk 1 above)
Comment (kevinbarabash) [reply .circleci/config.yml]: Sounds good to me.

**Comment by kevinbarabash** [comment]
> Upload built file and provide a preview/demo/sandbox for each build(PR and push)

We should probably only update the demo page whenever we publish.

**Comment by ylemkimon** [comment]
@kevinbarabash I agree the demo page should use latest release version.

However, it may be useful to provide the demo for old versions(git tags) like `https://khan.github.io/KaTeX/index.html?v=0.6.0` and preview for PRs like `https://khan.github.io/KaTeX/index.html?pr=1187`. And this can be done easily by using CircleCI build artifacts API.

**Comment by edemaine** [comment]
Yes, I think this would be cool, especially for testing PRs without having to grab the branch yourself.

**Comment by kevinbarabash** [comment]
> Run lint/flow/jest

I guess I should've looked at the PR a bit more careful.

**Comment by kevinbarabash** [comment]
> We'll definitely want this before merging. I guess I should've looked at the PR a bit more careful.

I guess we don't **need** it for this PR since we're still running travis-ci.  It's probably a good idea to run both for a while so that we have time to discover any rough edges circleci may have before deprecating travis-ci.

**Comment by kevinbarabash** [comment]
I like that we're able to run multiple jobs with different settings in parallel.

screen shot 2018-03-06 at 12 49 40 am

**Comment by kevinbarabash** [comment]
@ylemkimon thanks for the PR.

**Comment by edemaine** [comment]
I tried setting up my KaTeX fork on my CircleCI account, and it started building for the one PR that was on my fork.  I canceled that, and eventually found an interface to build my negspace fork (corresponding to #1194 here, which I rebased to include this PR).  Now I get [this error message](https://circleci.com/gh/edemaine/KaTeX/2):

```
* Cannot find a job named `build` to run in the `jobs:` section of your configuration file.
If you expected a workflow to run, check your config contains a top-level key called 'workflows:'
```

image

Am I doing something wrong?

**Comment by kevinbarabash** [comment]
@edemaine have you rebased your branch so that it includes the .circleci/config.yml?

**Comment by edemaine** [comment]
@kevinbarabash Yes.  I just double checked: https://github.com/edemaine/KaTeX/tree/negspace

**Comment by ylemkimon** [comment]
There seems to be similar issues reported: https://discuss.circleci.com/t/if-you-expected-a-workflow-to-run-check-your-config-contains-a-top-level-key-called-workflows/16798/6. Could you try pushing a new commit? I think we should mention this bug at our documentation.

**Comment by edemaine** [comment]
Cool, a dummy commit fixed it.

**Comment by ylemkimon** [comment]
@kevinbarabash After a week of using CircleCI, following problems were discovered:
1. Unlike Travis, CircleCI builds are tied to _commits_ pushed to the repo, not _merge commits_ generated by GitHub when a pull request is opened. As a result:
**a.** it tests the PR branch, not "will this work if I press the merge button?". So if the PR branch is out-of-date, the test result may differ.
This is worked around by either merging into the base branch before the test—this assumes there is only one PR per a branch—or constantly merging the base branch into the PR branch.
**b.** another PR with same HEAD but different base branch report previous build results.
**c.** if CircleCI is enabled on the repo where the PR branch is, it is built on that repo and its results are reported.
(**Related:** [Feature request: Show test results for prospective merge of a Github PR](https://discuss.circleci.com/t/show-test-results-for-prospective-merge-of-a-github-pr/1662))

**Comment by kevinbarabash** [comment]
@ylemkimon thanks for finding out all that info.  It's quite unfortunate about the circelci behavior.  :(

**Comment by edemaine** [comment]
This is not the *worst* news: we forbid merging PRs that are out-of-date with respect to the master branch, so we are already "constantly merging the base branch into the PR branch" (if I understand correctly).  And the resulting tests have to pass before merging.  So it's functional, though not ideal.

**Comment by ylemkimon** [comment]
@edemaine I thought `Update branch` button was just for convenience.
> Once you've set up required status checks, your branch must be up to date with the base branch before merging. This ensures that your branch has been tested with the latest code from the base branch. If your branch is out of date, you'll need to merge the base branch into your branch. ([GitHub documentation](https://help.github.com/articles/about-required-status-checks/))

Then, 1-a and 1-b won't be much of a problem.

**Pull request merged**